### ISSUE TITLE
Validate and parse kwargs when passed to struct constructor

### ISSUE THREAD

**Issue by Goldziher** [jcrist/msgspec#301]
### Description

Hi, so this is really a big problem for us - currently if I do this:

```python

class MyStruct(Struct):
    a: int

instance = MyStruct(a="not a number"). # should raise a validation error, but doesn't
```

There is no parsing and no validation happening at all. How would i go about this? `from_builtins` is meant to construct only from builtins and simpler types, and afaik it also doesnt perform the kind of parsing and validation we need. Thus, this:

```python
signature = from_builtins(
    kwargs,
    cls,
    dec_hook=partial(
        default_serializer,
        type_encoders={
            **DEFAULT_TYPE_ENCODERS,
            **connection.route_handler.resolve_type_encoders(),
            Request: lambda x: x,
            WebSocket: lambda x: x,
            State: lambda x: x,
            ASGIConnection: lambda x: x,
        },
    ),
)
```
Results in: `TypeError("from_builtins doesn't support objects of type 'WebSocket'")`

And this:

```python
signature = from_builtins(kwargs, cls, dec_hook=partial(default_serializer, type_encoders={**DEFAULT_TYPE_ENCODERS, **connection.route_handler.resolve_type_encoders()}), builtin_types=(
                Request, WebSocket, State, ASGIConnection
            ))
```

Results in: `TypeError("Cannot treat <class 'starlite.connection.request.Request'> as a builtin type")`

So, to be clear about our use case - we are not going to pass a simple json string, its not our use case, rather we have a dictionary of values composed from different data sources which need to be parsed and validated before being transformed into a dictionary of parsed values. We also use non-builtin custom types, as shown above, and the user can also define their own custom types etc + we support plugins. We therefore need some way to handle the above scenario.

**Comment by jcrist** [comment]
Can you provide a complete realistic example so we can talk more concretely about your use case? What is an example value of kwargs? What is the signature of `cls`? In this example where do the different parts of `kwargs` come from in the request, and what parts of that code path definitely require validation (i.e. contain user request-data)?

**Comment by Goldziher** [comment]
Sure, I'll send you a link to a PR with some failing tests that will demonstrate this.

**Comment by Goldziher** [comment]
Here @jcrist - this is how we will be using this: https://github.com/starlite-api/starlite/pull/1145/files#diff-6846ccb3c5d37139c1222fb2f99a55a8e933b097f68461865644f12eb038c78aR55, same as we use pydantic: https://github.com/starlite-api/starlite/pull/1145/files#diff-6846ccb3c5d37139c1222fb2f99a55a8e933b097f68461865644f12eb038c78aL296

The signature of `cls` is generated dynamically out of the signature of a route handling function. 

For example, lets say we have a "route handler" as its called in Starlite:

```python
@get("/{path_param:int}")
async def handler(path_param: int, query_params: list[int])
   ....
```

We will be generating a struct for this that would look like this:

```python
class SignatureModel(Struct):
   path_param: int
   query_params: list[int]
```

Of course, we would be doing this for any number of different signatures, many a lot more complex than the above examples.

The draft generation logic for the signature model you can see here:  https://github.com/starlite-api/starlite/pull/1145/files#diff-13784d3499f8ff682f42a418572a5caa62dca10542a3742d1db8785ebedbe037R201

The data the can be used in a route handler includes the following: path, query, header and cookie parameters. Request body values which can be JSON, msgpack, multipart/formdata and urlencoded/formdata, dependencies injected via dependency injection and framework classes (State, Request etc.). Additionally classes from other libraries via plugins, such as SQLAlchemy models can be used there. We inherently support dataclases, typedicts, named tuples and pydantic models, so these are common values as well.

Starlite handles the parsing of path parameters when parsing the url and matching the route. It handles parsing of query / header / cookie parameters into key/value string pairs, but does not do any type validation / parsing / coercion for these. For body values it handles the parsing from the encoded value into a python dictionary or list (depending on the content), but does not do this further. At present we rely on pydantic to handle the type convesion / coercion etc. 

So to summarize - we need to pass to the `SignatureModel` a dictionary of arbitrary values, which will be validated and the types will be correctly coerced as required. 

Another small example:

```python
class MyPydanticModel(BaseModel):
   value: int

@post("/")
def my_post_handler(body: MyPydanticModel) -> None:
    ....
```

The body kwarg in the above handler is typed as a pydantic model. We will therefore pass the request body dictionary to the pydantic model to instantiate it. This is a classic use case - we would like to also be able to support it using Struct types, as we do dataclasses, typed dicts etc.

**Comment by Goldziher** [comment]
@jcrist - don't want to nag you, but I would love to have the msgspec integration out as part of the Starlite v2 rework which we are slated to release at the end of the month. For this to happen though, we will need to resolve the above issue and tackle whatever other issues popup - if any. I think this is the last hurdle, but I cant be sure until we get past it.

**Comment by jcrist** [comment]
Thanks for the ping @Goldziher (and apologies for the delay)! I was heads down trying to get the 0.13.0 release out, then got distracted by other things. It's already late here, but I promise I'll give you a longer writeup response tomorrow.

**Comment by jcrist** [comment]
Thanks for your patience here. It sounds like a request handler has the following sources for parameters that can passed into it:
- Request URL (path components, query parameters)
- Request body (JSON, msgpack, formdata, ...)
- Starlite objects as requested (`Request`, `State`, ...)
- Other things from dependency injection/plugins

Right now you're translating a handler signature into a single pydantic model. Then on each request you merge all the data sources above into a big dictionary, and pass it to `Model(**kwargs)` to handle the parsing/coercion, before converting that back to a dictionary.

This works fine for pydantic, but `msgspec` intentionally doesn't support this pattern because it has some flaws:

- To handle parsing in this fashion, you need to have converted all your data sources (`url`, request body `json`/`msgpack`/...) into python objects _before_ handing them off to the validation library. `msgspec` handles both parsing and validation at the same time (for both efficiency and correctness). Forcing the model you're currently using will lose out on most of the performance benefits of `msgspec`.
- It also makes it harder to enforce "strict" parsing (to use pydantic's term). Query parameters are untyped strings, you'll always have to cast them to the requested type when parsing. JSON on the other hand has several types (`number`, `string`, ...). To support both in the same model means that any request sent expecting an integer JSON body parameter `a` and an integer query parameter `b` *may* end up allowing coercing a string `"1"` parameter for `a` rather than requiring a strict integer `1`. Since pydantic allows configuring strictness on a per-model basis you might not have this problem. But msgspec assumes `from_builtins` is used to wrap a _single_ protocol, meaning its coercion settings are global.
- It forces validation of things at the framework level you already know to be true. If a handler requests a `Request` argument from the framework, you don't really need to type validate that the `request` argument you pass in is a `Request`, since it's your code that's passing it in.

To work best with `msgspec`, I recommend the following pattern:

For each request handler, generate (as needed)

- A model for the request body
- A model for all path, query, cookie, and header parameters (assuming these all have the same coercion rules)
- Some code to handle the remaining dependency-injected arguments. Here I'm assuming everything else doesn't need validation, since it's only passed in if the user requests it. If it _does_ need validation in some form then I'll need more information on what you want here.

Then when handling the request you'd:

- Use either `msgspec` or `pydantic` to parse the request body directly into the body model. This lets `msgspec` efficiently handle `json` and `msgpack` for you without requiring unnecessary allocations.
- Use your normal parsing code to make a big dict of all the other info (path, query, cookie, and header parameters), then pass that to either pydantic or `msgspec.from_builtins` to parse the other model.
- Merge these two models into a single dict, along with the other dependency-injected arguments
- Call the handler function

This lets `msgspec` handle the parsing and validation tasks it was written to do, while also avoiding validating framework-internal code that shouldn't need validating anyway (IIUC). Does that make sense?

**Comment by Goldziher** [comment]
Hi @jcrist , and thanks for the answer. 

I considered this in advance to be honest, but it requires huge reworks of certain parts of our logic, and it does pose some challenges I'm terms of our API. I will try to do this and see how feasible it is - maybe my intuition is wrong. 

On an aaide- the lack of parsing and validation on struct init (constructor) makes it very difficult to work with msgspec as we do with pydantic. It will mean that users would not be able to use msgspec structs as utility classes, DTO objects, or part of our plugin system the same way pydantic models are used. It's not a blocker on our end, but it does heart adoption of the library - I think that a light, fast alternative to pydantic is really a killer.

**Comment by jcrist** [comment]
> On an aaide- the lack of parsing and validation on struct init (constructor) makes it very difficult to work with msgspec as we do with pydantic. It will mean that users would not be able to use msgspec structs as utility classes, DTO objects, or part of our plugin system the same way pydantic models are used.

I'm not a starlite user, so I'm not sure I follow. Why would those objects requiring parsing/validation as part of the constructor? Can you give an example?

**Comment by Goldziher** [comment]
> > On an aaide- the lack of parsing and validation on struct init (constructor) makes it very difficult to work with msgspec as we do with pydantic. It will mean that users would not be able to use msgspec structs as utility classes, DTO objects, or part of our plugin system the same way pydantic models are used.
> 
> I'm not a starlite user, so I'm not sure I follow. Why would those objects requiring parsing/validation as part of the constructor? Can you give an example?

This isn't about Starlite in particular, this is about the usage of pydantic. Even python's builtin dataclasses offer validation as part of the constructor call (but no parsing - this isnt a part of dataclasses in python). This is basically the main utility of both of these, with pydantic offering parsing as well. 

If you want this library to be widely used, aside from specialized use cases, than the main use cases of pydantic and dataclasses need to be central to it. These are - easy validation as part of a class constructor. Easy parsing in the same way. Easy conversion to and from other classes. etc. EASY is the keyword here.

Currently, the only way one can use your type parsing and validation is either to start from pythonic primitives, or start from JSON / msgpeck serialized data. This is very specialized and it basically gates the parsing and validation logic to the utilities. If this is the scope of this library and thats the aim, then thats perfectly fine of course, but if what you want is to have a wider appeal than you should consider basically offering an alternative to dataclasses and pydantic models.

**Comment by jcrist** [comment]
> Even python's builtin dataclasses offer validation as part of the constructor call 

I don't think that's true. Do you have a link for this? Or an example showing what you mean by this?

>  If this is the scope of this library and thats the aim, then thats perfectly fine of course, but if what you want is to have a wider appeal than you should consider basically offering an alternative to dataclasses and pydantic models.

`msgspec` is for:

- Fast encoding of python objects into common protocols 
- Fast decoding of common protocols into python objects, and validating the messages match a certain schema
- It also happens to provide a faster dataclass-like type (`msgspec.Struct`) with higher performance than dataclasses/pydantic on `__init__`/`__eq__`/`__repr__`/`__copy__`/...

We deliberately do not do anything fancy like parsing/validation in the `__init__` of struct types. This is for the following reasons:

- Python's dataclasses don't, neither do most builtin types (`list`, `dict`, ...). Doing validation on `__init__` of `Struct` types alone won't catch all bugs, and leads to discrepancies between all the different runtime types a user might use.
- Even if we did validate on `__init__`, the user can always modify the struct later on after `__init__`. Do we also validate on `__setattr__`? Even that won't catch everything; we can't validate if the user modifies a `list` already set on a struct.
- Validating in an `__init__` leads to poorer performance, especially if the Struct takes nested objects (e.g. a list of lists of ints). In certain uses this can lead to scanning the same list multiple times just to validate nested constructors!
- When is `__init__` called? Is it called when you do a `__copy__`? What about when unpickling a value?

I'd rather be explicit about when validation happens. In msgspec it happens at the boundaries of your program, when processing input values from some arbitrary source. It doesn't happen on internal usage, since IMO validating the correctness of your program internals is better handled by type checkers, linters, and unit tests.

In short:

- `msgspec` is for parsing and validating input data to your application, which will always need some runtime checks.
- It's not for validating the internals of your application, which are better handled by type checkers and unit tests.

Here input data may be:

- Data in one of the supported formats (JSON, msgpack, TOML, or YAML). These have builtin modules (`msgspec.json`) to handle the parsing. I'm not sure how things could get easier than that.
- Data in an unsupported format (e.g. CBOR or BSON). These may be handled by wrapping another serialization library ( which generally only work with builtin python types) with `from_builtins`/`to_builtins`.
- Data coming from a database/object store/... These again generally only return builtin python types, and so may be coerced and validated to the proper type using `from_builtins`.

If it helps you may think of it like [go's JSON](https://pkg.go.dev/encoding/json) or [Rust's Serde](https://serde.rs/) - a tool for (de)serialization and validation to specific object structures. `msgspec` also includes some support for [constraints](https://jcristharif.com/msgspec/constraints.html), which neither of those tools support out of the box. And yet both are still widely used.

**Comment by jcrist** [comment]
Based on conversation in #325 I believe this can be closed.
